journal/src/test/java/io/zeebe/journal/file/SegmentedJournalTest.java

```diff
@@ -0,0 +1,376 @@
+
+    int writerIndex;
+    for (writerIndex = 1; writerIndex <= totalWrites; writerIndex++) {
+      final var record = journal.append(asqn++, data);
+      assertThat(record.index()).isEqualTo(writerIndex);
+      written.put(writerIndex, record);
+    }
+
+    int readerIndex;
+    for (readerIndex = 1; readerIndex <= truncateIndex; readerIndex++) {
+      assertThat(reader.hasNext()).isTrue();
+      final var record = reader.next();
+      assertThat(record.index()).isEqualTo(readerIndex);
+      assertThat(record.asqn()).isEqualTo(written.get(readerIndex).asqn());
+    }
+
+    // when
+    journal.deleteAfter(truncateIndex);
+
+    for (writerIndex = truncateIndex + 1; writerIndex <= totalWrites; writerIndex++) {
+      final var record = journal.append(asqn++, data);
+      assertThat(record.index()).isEqualTo(writerIndex);
+      written.put(writerIndex, record);
+    }
+
+    for (; readerIndex <= totalWrites; readerIndex++) {
+      assertThat(reader.hasNext()).isTrue();
+      final var record = reader.next();
+      assertThat(record.index()).isEqualTo(readerIndex);
+      assertThat(record.asqn()).isEqualTo(written.get(readerIndex).asqn());
+    }
+  }
+
+  @Test
+  public void shouldReadAfterCompact() {
+    // given
+    final var reader = journal.openReader();
+    long asqn = 1;
+
+    for (int i = 1; i <= entriesPerSegment * 5; i++) {
+      assertThat(journal.append(asqn++, data).index()).isEqualTo(i);
+    }
+    assertThat(reader.hasNext()).isTrue();
+
+    // when - compact up to the first index of segment 3
+    final int indexToCompact = entriesPerSegment * 2 + 1;
+    journal.deleteUntil(indexToCompact);
+
+    // then
+    assertThat(reader.hasNext()).isTrue();
+    assertThat(reader.next().index()).isEqualTo(indexToCompact);
+  }
+
+  @Test
+  public void shouldSeekToIndex() {
+    // given
+    final var reader = journal.openReader();
+    long asqn = 1;
+    JournalRecord lastRecordWritten = null;
+    for (int i = 1; i <= entriesPerSegment * 2; i++) {
+      final JournalRecord record = journal.append(asqn++, data);
+      assertThat(record.index()).isEqualTo(i);
+      lastRecordWritten = record;
+    }
+    assertThat(reader.hasNext()).isTrue();
+
+    // when - compact up to the first index of segment 3
+    reader.seek(lastRecordWritten.index() - 1);
+
+    // then
+    assertThat(reader.hasNext()).isTrue();
+    assertThat(reader.next().index()).isEqualTo(lastRecordWritten.index() - 1);
+  }
+
+  @Test
+  public void shouldSeekToAsqn() {
+    // given
+    final var reader = journal.openReader();
+    long asqn = 10;
+    JournalRecord lastRecordWritten = null;
+    for (int i = 1; i <= entriesPerSegment * 2; i++) {
+      final JournalRecord record = journal.append(asqn++, data);
+      assertThat(record.index()).isEqualTo(i);
+      lastRecordWritten = record;
+    }
+    assertThat(reader.hasNext()).isTrue();
+
+    // when
+    reader.seekToAsqn(lastRecordWritten.asqn() - 2);
```

Comment:
why minus 2 ? :sweat_smile:

Reply:
Just randomly chosen. :slightly_smiling_face: Wanted to test that it can seek to record which is not the first or the last one. There should be more than one tests to cover specific cases - seek to first record, seek to last one, seek to non-existing ones etc. which should be done as part of #6222